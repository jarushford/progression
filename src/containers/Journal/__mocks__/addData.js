export const addDataThunk = jest.fn()