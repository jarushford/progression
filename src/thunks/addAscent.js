import { fetchAscentsThunk } from './fetchAscents'
import { clearAscents, setError } from '../actions'

export const addAscentThunk = (ascent) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/ascents/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(ascent),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add ascent')
      }
      await dispatch(clearAscents())
      dispatch(fetchAscentsThunk(ascent.user_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}