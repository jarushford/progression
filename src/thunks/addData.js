import { clearThunkHelper } from '../utils/thunkHelpers'
import { fetchDataThunk } from './fetchData'
import { setError } from '../actions'

export const addDataThunk = (data, type) => {
  return async (dispatch) => {
    try {
      const url = `http://localhost:3000/api/progressionusers/${type}s/new`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(data),
        headers: {
          "Content-Type": "application/json"
        }
      })

      if (!response.ok) {
        throw Error(`Could not add ${type}`)
      }

      await dispatch(clearThunkHelper(type)())
      type === 'journal' || type === 'milestone'
        ? dispatch(fetchDataThunk(data.user_id, data.project_id, type))
        : dispatch(fetchDataThunk(data.user_id, '', type))

      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}