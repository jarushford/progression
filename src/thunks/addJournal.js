import { fetchJournalThunk } from './fetchJournal'
import { clearJournal, setError } from '../actions'

export const addJournalThunk = (entry) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/journal/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(entry),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add entry')
      }
      await dispatch(clearJournal())
      dispatch(fetchJournalThunk(entry.user_id, entry.project_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}