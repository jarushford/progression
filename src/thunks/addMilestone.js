import { fetchMilestonesThunk } from './fetchMilestones'
import { clearMilestones, setError } from '../actions'

export const addMilestoneThunk = (milestone) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/milestones/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(milestone),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add project')
      }
      await dispatch(clearMilestones())
      dispatch(fetchMilestonesThunk(milestone.user_id, milestone.project_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}