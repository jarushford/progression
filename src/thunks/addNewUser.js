import { setError } from '../actions'
import { loginUserThunk } from './loginUser'

export const addNewUserThunk = (user) => {
  return async (dispatch) => {
    try {
      const url = `https://progression-backend.herokuapp.com/api/progressionusers/new`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(user),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw new Error('User already exists, try another email!')
      }
      await dispatch(loginUserThunk(user))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}