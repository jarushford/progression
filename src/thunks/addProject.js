import { fetchProjectsThunk } from './fetchProjects'
import { clearProjects } from '../actions'

export const addProjectThunk = (project) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/projects/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(project),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add project')
      }
      await dispatch(clearProjects())
      dispatch(fetchProjectsThunk(project.user_id))
    } catch (error) {
      console.log(error)
    }
  }
}