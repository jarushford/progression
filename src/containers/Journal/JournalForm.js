import React, { Component } from 'react'
import { connect } from 'react-redux'
import { addDataThunk } from '../../thunks/addData'
import { Redirect } from 'react-router-dom'

class JournalForm extends Component {
  constructor() {
    super()
    this.state = {
      entry: '',
      entryAdded: false
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({ [name]: value })
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    const { user, project } = this.props
    const date = new Date()
    const today = 
    `${('0' + (date.getMonth() + 1)).slice(-2)}/`
      + `${('0' + (date.getDate())).slice(-2)}/`
      + `${date.getFullYear()}`
    const entry = {
      entry: this.state.entry,
      journal_date: today,
      user_id: user.id,
      project_id: project
    }
    const result = await this.props.addJournal(entry, 'journal')
    if (result) {
      this.setState({ entryAdded: true })
    }
  }

  render() {
    const { entry, entryAdded } = this.state

    if (entryAdded) { return <Redirect to={`/projects/${this.props.project}`} /> }

    return (
      <form className="journal-form" onSubmit={this.handleSubmit}>
        <h2>ADD JOURNAL ENTRY</h2>
        <textarea name="entry"
          value={entry}
          onChange={this.handleChange}
          placeholder="entry"
          maxLength="800"
          className="journal-entry"
        />
        <button>Add Entry</button>
      </form>
    )
  }
}

const mapStateToProps = (state) => ({
  user: state.currentUser,
  project: state.currentProject
})

const mapDispatchToProps = (dispatch) => ({
  addJournal: (entry, type) => dispatch(addDataThunk(entry, type))
})

export default connect(mapStateToProps, mapDispatchToProps)(JournalForm)

