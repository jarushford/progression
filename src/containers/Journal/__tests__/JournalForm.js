import React from 'react'
import { shallow } from 'enzyme'
import { addDataThunk } from '../../../thunks/addData'
import { mapStateToProps, mapDispatchToProps, JournalForm } from '../JournalForm'

jest.mock('../../../thunks/addData')

describe('Journal Form', () => {
  let mockAddEntry
  let mockUser
  let mockProject

  beforeEach(() => {
    mockUser = {
      name: 'T-Rex',
      email: 'chompy@yahoo.com',
      password: 'smallarms',
      id: 4
    }
    mockProject = 2
    mockAddEntry = jest.fn().mockImplementation(() => true)
  })

  describe('JournalForm Component', () => {
    it('should match the snapshot', () => {
      const wrapper = shallow(<JournalForm addJournal={mockAddEntry} user={mockUser} project={mockProject} />)

      expect(wrapper).toMatchSnapshot()
    })

    it('should have the correct default state', () => {
      const expected = { entry: '', entryAdded: false }
      const wrapper = shallow(<JournalForm addJournal={mockAddEntry} user={mockUser} project={mockProject} />)

      expect(wrapper.state()).toEqual(expected)
    })

    it('should update state on input change', () => {
      const expected = 'Rocks!'
      const e = { target: { name: 'entry', value: 'Rocks!' } }
      const wrapper = shallow(<JournalForm addJournal={mockAddEntry} user={mockUser} project={mockProject} />)

      wrapper.find('textarea').simulate('change', e)

      expect(wrapper.state().entry).toEqual(expected)
    })

    it('should add a journal entry on submit', async () => {
      const wrapper = shallow(<JournalForm addJournal={mockAddEntry} user={mockUser} project={mockProject} />)
      wrapper.setState({ entry: 'Rocks!' })
      const e = { preventDefault: jest.fn() }

      await wrapper.find('.journal-form').simulate('submit', e)

      expect(wrapper.state().entryAdded).toEqual(true)
    })

    it('should not reset state if the entry is not added', async () => {
      mockAddEntry.mockImplementation(() => false)
      const wrapper = shallow(<JournalForm addJournal={mockAddEntry} user={mockUser} project={mockProject} />)
      wrapper.setState({ entry: 'Rocks!' })
      const e = { preventDefault: jest.fn() }

      await wrapper.find('.journal-form').simulate('submit', e)

      expect(wrapper.state().entryAdded).toEqual(false)
    })
  })

  describe('mapStateToProps', () => {
    it('should return a props object with data from store', () => {
      const mockState = { currentUser: mockUser, currentProject: mockProject }
      const expected = { user: mockUser, project: mockProject }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should return a props object with a method addJournal', () => {
      const mockDispatch = jest.fn()
      const expected = addDataThunk()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.addJournal()

      expect(mockDispatch).toBeCalledWith(expected)
    })
  })
})