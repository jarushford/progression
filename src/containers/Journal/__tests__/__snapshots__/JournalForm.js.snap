// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Journal Form JournalForm Component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <JournalForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          ADD JOURNAL ENTRY
        </h2>,
        <textarea
          className="journal-entry"
          maxLength="800"
          name="entry"
          onChange={[Function]}
          placeholder="entry"
          value=""
        />,
        <button>
          Add Entry
        </button>,
      ],
      "className": "journal-form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "ADD JOURNAL ENTRY",
        },
        "ref": null,
        "rendered": "ADD JOURNAL ENTRY",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "journal-entry",
          "maxLength": "800",
          "name": "entry",
          "onChange": [Function],
          "placeholder": "entry",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add Entry",
        },
        "ref": null,
        "rendered": "Add Entry",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            ADD JOURNAL ENTRY
          </h2>,
          <textarea
            className="journal-entry"
            maxLength="800"
            name="entry"
            onChange={[Function]}
            placeholder="entry"
            value=""
          />,
          <button>
            Add Entry
          </button>,
        ],
        "className": "journal-form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "ADD JOURNAL ENTRY",
          },
          "ref": null,
          "rendered": "ADD JOURNAL ENTRY",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "journal-entry",
            "maxLength": "800",
            "name": "entry",
            "onChange": [Function],
            "placeholder": "entry",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Entry",
          },
          "ref": null,
          "rendered": "Add Entry",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
